engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/command/PartiallyOrderedAsynchronousCommand.scala

```diff
@@ -0,0 +1,91 @@
+package org.enso.interpreter.instrument.command
+
+import com.oracle.truffle.api.TruffleLogger
+import org.enso.interpreter.instrument.execution.{Completion, RuntimeContext}
+import org.enso.interpreter.instrument.execution.Completion.{Done, Interrupted}
+import org.enso.interpreter.runtime.control.ThreadInterruptedException
+import org.enso.polyglot.runtime.Runtime.Api.RequestId
+
+import java.util.logging.Level
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.control.NonFatal
+import scala.util.{Failure, Success}
+
+/** `PartiallyOrderedAsynchronousCommand` will execute asynchronously along with other commands except that any
+  * partially ordered commands will always execute in the same order they were submitted.
+  *
+  * The class was added since some commands cannot be executed in arbitrary order but they also must not be
+  * executed synchronously due to the possibility of deadlocks.
+  * Plain context locks do not necessarily guarantee the right order of such commands.
+  */
+abstract class PartiallyOrderedAsynchronousCommand(
+  maybeRequestId: Option[RequestId]
+) extends Command(maybeRequestId) {
+
+  final override def execute(implicit
+    ctx: RuntimeContext,
+    ec: ExecutionContext
+  ): Future[Completion] = {
+    val logger = ctx.executionService.getLogger
+    val idx    = ctx.locking.getUniqueTaskIndex()
+    for {
+      _ <- Future {
+        logger.log(Level.FINE, s"Executing command asynchronously: $this...")
+      }
+      result <- mapFailures(
+        logger,
+        executePartiallyOrdered(ctx, ec, idx)
+      )
+      _ <- Future {
+        logger.log(Level.FINE, s"Command $this finished.")
+      }
+    } yield result
+  }
+
+  private def mapFailures(logger: TruffleLogger, result: Future[Unit])(implicit
+    ex: ExecutionContext
+  ): Future[Completion] = {
+    result.transformWith[Completion] {
+      case Success(()) =>
+        Future.successful(Done)
+
+      case Failure(_: InterruptedException | _: ThreadInterruptedException) =>
+        Future.successful[Completion](Interrupted)
+
+      case Failure(NonFatal(ex)) =>
+        logger.log(
+          Level.SEVERE,
+          s"An error occurred during execution of $this command",
+          ex
+        )
+        Future.failed[Completion](ex)
+
+      case Failure(ex) =>
+        logger.log(
+          Level.SEVERE,
+          s"An error occurred during execution of $this command",
+          ex
+        )
+        Future.failed[Completion](ex)
+    }
+  }
+
+  private def executePartiallyOrdered(implicit
+    ctx: RuntimeContext,
+    ec: ExecutionContext,
+    idx: Int
+  ): Future[Unit] = {
+    ctx.locking.waitOnTaskIndex(idx)
+    try {
+      executeAsynchronously
+    } finally {
+      ctx.locking.advanceTaskIndex()
+    }
```

Comment:
Is this surely correct?

Inside of this `try` block, we run `executeAsynchronously` which prepares, starts and returns a `Future[Unit]`. We then enter the `finally` block and advance the task index, but we don't know if the `Future` that we are about to return has finished execution.


---

Do we have some tests that check the synchronization of this? e.g. for OpenFileCmd etc.

I think this should be tested. It's not a trivial piece of logic. We have infrastructure for testing parallelism and synchronization, see for example `TestSynchronizer` and its usages.